spmd/tensor/ops/tp_sharding_ops.py

```diff
@@ -4,40 +4,36 @@
 import torch.utils._pytree as pytree
 from typing import List
 from spmd.tensor.api import DTensor
+from spmd.tensor.dispatch import OpSchema, OutputSharding
 from spmd.tensor.utils import unwrap_local_tensor
-from spmd.tensor.ops.utils import unwrap_single_placement, register_impl
+from spmd.tensor.ops.utils import unwrap_single_placement, register_impl, register_prop_rule
+from spmd.tensor.placement_types import DTensorSpec
 
 """
 The ops below were quickly hacked and needed to be polished down the road.
 Although they come with unit tests already, the logic is directly borrowed
-from ShardedTensor. We need to also make it work for all placement types 
+from ShardedTensor. We need to also make it work for all placement types
 of DTensor and all corner cases for sharded distributed tensor.
 """
 
 
-@register_impl("aten.baddbmm.default")
-def dist_baddbmm(
-    self: DTensor,
-    batch1: DTensor,
-    batch2: DTensor,
-    beta: float = 1.0,
-    alpha: float = 1.0,
-) -> DTensor:
-    local_input, local_batch1, local_batch2 = pytree.tree_map(
-        unwrap_local_tensor, (self, batch1, batch2)
+@register_prop_rule("aten.bmm.default")
+def bmm_rules(op_schema: OpSchema) -> OutputSharding:
+    mat1_spec, mat2_spec = op_schema.args_spec
+    mat1_shape = list(mat1_spec.shape) # b * n * m
```

Comment:
are these rules only aim to support sharding on the batch dimension? bc i don't see any propagation base on the rule written here

Reply:
According to this code:
https://github.com/pytorch/PiPPy/pull/382/files#diff-44be2c8678cd9e83382ce3e5d09f972b06935f8980bc9410cce9cac0651552d0R143-R150
I believe the current `bmm` only aims to support sharding on the batch dim. But I need to talk to @fduwjj to clarify on this.

> bc i don't see any propagation base on the rule written here
Right, my intention is to have default sharding prop on bmm (i.e. sharding on batch dim for both tensors, the result then has the same sharding). Or, I misunderstand the right approach to do it and my implementation won't do that?

Reply:
I think it's fine to only support sharding on batch dim for `bmm`, at least initially. If so, could you check the input specs, and if the input specs not sharding on the batch dim, return `OutputSharding.output_spec` as None, and add failed reason to `OutputSharding.failed_reson`?

Reply:
Thx! Will do it.  One question is, since Junjie wants to support the case where one matrix/tensor is replicated and the other is sharded, can I achieve it using DTensorSpec or do I have to manually redistribute the DTensor object like what's done here?
https://github.com/pytorch/PiPPy/pull/382/files#diff-44be2c8678cd9e83382ce3e5d09f972b06935f8980bc9410cce9cac0651552d0R146-R150

Reply:
You don't need to manually do reshard, what you could actually do is, if input is replicate + sharded, then it's good to run it locally, but if input is not either `replicate + sharded` and `sharded + sharded`, you can set output.spec to None, and set schema suggestions to the suggestion that you are expecting. Then the operator dispatch could automatically do this for you https://github.com/pytorch/PiPPy/blob/main/spmd/tensor/dispatch.py#L92